File: lib/src/test/java/com/hammwerk/wizardpager/core/PageTest.java
package com.hammwerk.wizardpager.core;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import de.bechte.junit.runners.context.HierarchicalContextRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@RunWith(HierarchicalContextRunner.class)
public class PageTest {
	public class GivenAPage {
		private Page page;

		@Before
		public void givenAPage() throws Exception {
			page = new TestPage("Page");
		}

		@Test
		public void titleShouldBeSetProperly() throws Exception {
			assertEquals("Page", page.getTitle());
		}

		@Test
		public void whenAskForValidity_thenReturnTrue() throws Exception {
			assertTrue(page.isValid());
		}

		public class GivenAPageListener {
			private PageListener pageListener;

			@Before
			public void givenAPageListener() throws Exception {
				pageListener = mock(PageListener.class);
				page.setPageListener(pageListener);
			}

			@Test
			public void whenSetCompleted_thenDontCallOnPageValidCallback() throws Exception {
				page.setCompleted();
				verify(pageListener, times(0)).onPageValid(page);
			}

			@Test
			public void whenSetNotCompleted_thenDontCallOnPageInvalidCallback() throws Exception {
				page.setNotCompleted();
				verify(pageListener, times(0)).onPageInvalid(page);
			}
		}
	}

	public class GivenARequiredPage {
		private TestPage requiredPage;

		@Before
		public void givenRequiredPage() throws Exception {
			requiredPage = new TestPage("Required Page", true);
		}

		@Test
		public void requiredPageShouldBeInvalidWhenCreated() throws Exception {
			assertFalse(requiredPage.isValid());
		}

		@Test
		public void whenSetCompleted_thenPageIsValid() throws Exception {
			requiredPage.setCompleted();
			assertTrue(requiredPage.isValid());
		}

		@Test
		public void whenSetCompletedAfterItWasCompleted_thenPageIsInvalid() throws Exception {
			requiredPage.setCompleted();
			requiredPage.setNotCompleted();
			assertFalse(requiredPage.isValid());
		}

		public class GivenAPageListener {
			private PageListener pageListener;

			@Before
			public void givenAPageListener() throws Exception {
				pageListener = mock(PageListener.class);
				requiredPage.setPageListener(pageListener);
			}

			@Test
			public void whenSetCompleted_thenCallOnPageValidCallback() throws Exception {
				requiredPage.setCompleted();
				verify(pageListener, times(1)).onPageValid(requiredPage);
			}

			@Test
			public void whenSetNotCompleted_thenCallOnPageInvalidCallback() throws Exception {
				requiredPage.setNotCompleted();
				verify(pageListener, times(1)).onPageInvalid(requiredPage);
			}
		}
	}
}

File: lib/src/test/kotlin/com/hammwerk/wizardpager/core/PageTest.kt
package com.hammwerk.wizardpager.core

import de.bechte.junit.runners.context.HierarchicalContextRunner
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.never
import org.mockito.Mockito.verify

@RunWith(HierarchicalContextRunner::class)
class PageTest {
	inner class GivenAPage {
		private lateinit var page: Page

		@Before
		fun setUp() {
			page = TestPage("Page")
		}

		@Test
		fun whenGetTitle_ThenReturnTitle() {
			assertEquals("Page", page.title)
		}

		@Test
		fun whenGetFragmentName_ThenReturnFragmentName() {
			assertEquals("FragmentName", page.fragmentName)
		}

		@Test
		fun whenAskForValidity_thenReturnTrue() {
			assertTrue(page.valid)
		}

		inner class GivenAnOnPageValidCallback {
			@Before
			fun setUp() {
				page.onPageValid = mock()
			}

			@Test
			fun whenSetCompleted_thenDontCallOnPageValidCallback() {
				page.completed = true
				verify(page.onPageValid, never())!!.invoke(page)
			}
		}

		inner class GivenAnOnPageInvalidCallback {
			@Before
			fun setUp() {
				page.onPageInvalid = mock()
			}

			@Test
			fun whenSetNotCompleted_thenDontCallOnPageInvalidCallback() {
				page.completed = false
				verify(page.onPageInvalid, never())!!.invoke(page)
			}
		}
	}

	inner class GivenARequiredPage {
		private lateinit var requiredPage: Page

		@Before
		fun setUp() {
			requiredPage = TestPage("Required Page", true)
		}

		@Test
		fun requiredPageShouldBeInvalidWhenCreated() {
			assertFalse(requiredPage.valid)
		}

		@Test
		fun whenSetCompleted_thenPageIsValid() {
			requiredPage.completed = true
			assertTrue(requiredPage.valid)
		}

		@Test
		fun whenSetNotCompletedAfterItWasCompleted_thenPageIsInvalid() {
			requiredPage.completed = true
			requiredPage.completed = false
			assertFalse(requiredPage.valid)
		}

		inner class GivenAnOnPageValidCallback {
			@Before
			fun setUp() {
				requiredPage.onPageValid = mock()
			}

			@Test
			fun whenSetCompleted_thenCallOnPageValidCallback() {
				requiredPage.completed = true
				verify(requiredPage.onPageValid)!!.invoke(requiredPage)
			}
		}

		inner class GivenAnOnPageInvalidCallback {
			@Before
			fun setUp() {
				requiredPage.onPageInvalid = mock()
			}

			@Test
			fun whenSetNotCompleted_thenCallOnPageInvalidCallback() {
				requiredPage.completed = false
				verify(requiredPage.onPageInvalid)!!.invoke(requiredPage)
			}
		}
	}
}